src/db/gc/mod.rs

```diff
@@ -0,0 +1,323 @@
+//! reachable, making sure they are kept in the database after the run. It makes sure to retain the
+//! reachable graph as well as all the blocks for at least `chain finality` to account for potential
+//! forks. A snapshot can be used to bootstrap the node from scratch, thus the algorithm is
+//! considered correct when a valid snapshot can be exported using records available in the database
+//! after the run.
+//!
+//! ## Disk usage
+//! There's no additional disk space required to run this algorithm. However, removing the
+//! unreachable blocks from the database takes at least `chain finality`. The GC speed depends on
+//! the reachable graph size. Disk usage is also affected by the GC run interval. Additionally,
+//! since a truncated 4-byte hash is used - there's a slight possibility of a collision, which might
+//! result in an unreachable block being retained in the database. Still, the impact on the total
+//! disk size is negligible.
+//!
+//! ## Memory usage
+//! During the `mark` and up to the `sweep` stage, the algorithm requires `4 bytes` of memory for
+//! each database record. Additionally, the seen cache while traversing the reachable graph
+//! executing the `filter` stage requires at least `32 bytes` of memory for each reachable block.
+//!
+//! ## Scheduling
+//! 1. GC is triggered automatically and there have to be at least `chain finality` epochs stored
+//! for the `mark` step.
+//! 2. The `filter` step is triggered after at least `chain finality` has passed since the `mark`
+//! step.
+//! 3. Then, the `sweep` step happens.
+//! 4. Finally, the algorithm waits for a configured amount of time to initiate the next run.
+//!
+//! ## Performance
+//! Note: This needs to be measured and potentially defined in terms of `snapshot export` or any
+//! other visible and comparable metric.
+use crate::blocks::Tipset;
+use crate::chain::ChainEpochDelta;
+
+use crate::db::{truncated_hash, GarbageCollectable};
+use crate::ipld::stream_graph;
+use crate::shim::clock::ChainEpoch;
+use ahash::{HashSet, HashSetExt};
+use futures::StreamExt;
+use fvm_ipld_blockstore::Blockstore;
+use std::mem;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time;
+use tracing::info;
+
+// This enum facilitates GC loop control flow. It allows for simpler workflow logic by avoiding
+// inner loops within the workflow itself.
+#[derive(PartialEq, Debug)]
+enum ControlFlow {
+    // Continue the GC workflow.
+    Continue,
+    // The GC workflow has finished.
+    Finished,
+}
+
+/// [`MarkAndSweep`] is a simple garbage collector implementation that traverses all the database
+/// keys writing them to a [`HashSet`], then filters out those that need to be kept and schedules
+/// the rest for removal.
+///
+/// Note: The GC does not know anything about the hybrid CAR-backed and ParityDB approach, only
+/// taking care of the latter.
+pub struct MarkAndSweep<DB, GT> {
+    db: Arc<DB>,
+    get_heaviest_tipset: GT,
```

Comment:
How about `dyn Fn() -> Arc<Tipset>`?

Reply:
I really don't want to box it and see no reason to. Why?

Reply:
It would be less code. Less is more. :)

Reply:
Let me see what comes out of it.